# Creacion, entrenamiento y despliegue de Modelo de ML

### Load_Digits Dataset (Optical Recognition of Handwritten Digits)

* Dataset load_digits API: https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_digits.html#sklearn.datasets.load_digits
* Dataset load_digits guide: https://archive.ics.uci.edu/dataset/80/optical+recognition+of+handwritten+digits

In [1]:
# Libraries
import warnings
from sklearn.datasets import load_digits

warnings.filterwarnings("ignore")

In [2]:
dataset = load_digits(as_frame=True)

In [3]:
dataset.keys()

dict_keys(['data', 'target', 'frame', 'feature_names', 'target_names', 'images', 'DESCR'])

In [4]:
dataset['data']

pixel_0_0  pixel_0_1  pixel_0_2  pixel_0_3  pixel_0_4  pixel_0_5  \
0           0.0        0.0        5.0       13.0        9.0        1.0   
1           0.0        0.0        0.0       12.0       13.0        5.0   
2           0.0        0.0        0.0        4.0       15.0       12.0   
3           0.0        0.0        7.0       15.0       13.0        1.0   
4           0.0        0.0        0.0        1.0       11.0        0.0   
...         ...        ...        ...        ...        ...        ...   
1792        0.0        0.0        4.0       10.0       13.0        6.0   
1793        0.0        0.0        6.0       16.0       13.0       11.0   
1794        0.0        0.0        1.0       11.0       15.0        1.0   
1795        0.0        0.0        2.0       10.0        7.0        0.0   
1796        0.0        0.0       10.0       14.0        8.0        1.0   

      pixel_0_6  pixel_0_7  pixel_1_0  pixel_1_1  ...  pixel_6_6  pixel_6_7  \
0           0.0        0.0        0.0        0.0  ...        0.0        0.0   
1           0.0        0.0        0.0        0.0  ...        0.0        0.0   
2           0.0        0.0        0.0        0.0  ...        5.0        0.0   
3           0.0        0.0        0.0        8.0  ...        9.0        0.0   
4           0.0        0.0        0.0        0.0  ...        0.0        0.0   
...         ...        ...        ...        ...  ...        ...        ...   
1792        0.0        0.0        0.0        1.0  ...        4.0        0.0   
1793        1.0        0.0        0.0        0.0  ...        1.0        0.0   
1794        0.0        0.0        0.0        0.0  ...        0.0        0.0   
1795        0.0        0.0        0.0        0.0  ...        2.0        0.0   
1796        0.0        0.0        0.0        2.0  ...        8.0        0.0   

      pixel_7_0  pixel_7_1  pixel_7_2  pixel_7_3  pixel_7_4  pixel_7_5  \
0           0.0        0.0        6.0       13.0       10.0        0.0   
1           0.0        0.0        0.0       11.0       16.0       10.0   
2           0.0        0.0        0.0        3.0       11.0       16.0   
3           0.0        0.0        7.0       13.0       13.0        9.0   
4           0.0        0.0        0.0        2.0       16.0        4.0   
...         ...        ...        ...        ...        ...        ...   
1792        0.0        0.0        2.0       14.0       15.0        9.0   
1793        0.0        0.0        6.0       16.0       14.0        6.0   
1794        0.0        0.0        2.0        9.0       13.0        6.0   
1795        0.0        0.0        5.0       12.0       16.0       12.0   
1796        0.0        1.0        8.0       12.0       14.0       12.0   

      pixel_7_6  pixel_7_7  
0           0.0        0.0  
1           0.0        0.0  
2           9.0        0.0  
3           0.0        0.0  
4           0.0        0.0  
...         ...        ...  
1792        0.0        0.0  
1793        0.0        0.0  
1794        0.0        0.0  
1795        0.0        0.0  
1796        1.0        0.0  

[1797 rows x 64 columns]

In [5]:
# Load the data into X and y
# X is the data and y is the target
X = dataset['data']
y = dataset['target']

X.shape, y.shape

((1797, 64), (1797,))

## Modelo 1 - Logistic Regression

[Scikit Learn - LogisticRegression](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html)

In [6]:
from sklearn.linear_model import LogisticRegression

In [7]:
model = LogisticRegression()
# Get the default parameters
model.get_params()

{'C': 1.0,
 'class_weight': None,
 'dual': False,
 'fit_intercept': True,
 'intercept_scaling': 1,
 'l1_ratio': None,
 'max_iter': 100,
 'multi_class': 'deprecated',
 'n_jobs': None,
 'penalty': 'l2',
 'random_state': None,
 'solver': 'lbfgs',
 'tol': 0.0001,
 'verbose': 0,
 'warm_start': False}

In [8]:
# Fit the model
# The model is fit to the data
# The model learns the relationship between X and y
# The model is trained on the data
model.fit(X, y)

LogisticRegression()

In [9]:
# Display the coefficients
# The coefficients are the weights of the model
# The coefficients are the values that the model learned
# The coefficients are used to make predictions
model.coef_

array([[ 0.00000000e+00, -5.45643843e-03, -4.12130745e-02,
         9.47979190e-02,  1.14111634e-02, -9.95292924e-02,
        -1.21942291e-01, -1.19397761e-02, -1.53988927e-04,
        -5.57845896e-02,  3.03414665e-02,  1.54087261e-01,
         8.81995241e-02,  1.30807851e-01, -5.13494588e-02,
        -1.20055567e-02, -1.72062022e-04,  4.15469140e-02,
         2.00672409e-01, -3.70270089e-02, -3.59401940e-01,
         1.96910240e-01,  8.57320712e-02, -4.12225049e-03,
        -7.83365223e-05,  1.53693154e-01,  1.26112095e-01,
        -2.41707620e-01, -5.44794696e-01,  2.87935599e-02,
         9.53895438e-02, -2.73015196e-04,  0.00000000e+00,
         1.87455501e-01,  1.20603629e-01, -2.52209638e-01,
        -4.77603514e-01,  4.12647203e-02,  6.75424465e-02,
         0.00000000e+00, -6.70032943e-04, -2.59723039e-02,
         3.00166268e-01, -1.75567748e-01, -1.57136766e-01,
         1.13928145e-01,  3.51638304e-02, -7.55014366e-04,
        -1.90357874e-03, -1.00859842e-01,  1.23677952e-0

In [10]:
predict1 = model.predict(X)

In [11]:
for i in enumerate(predict1):
    # Print the predicted value and the actual value
    print(f"Predicted: {i[1]}, Actual: {y[i[0]]}")
    if i[1] != y[i[0]]:
        print(f"  Mismatch: {i[1]} != {y[i[0]]}")

Predicted: 0, Actual: 0
Predicted: 1, Actual: 1
Predicted: 2, Actual: 2
Predicted: 3, Actual: 3
Predicted: 4, Actual: 4
Predicted: 5, Actual: 5
Predicted: 6, Actual: 6
Predicted: 7, Actual: 7
Predicted: 8, Actual: 8
Predicted: 9, Actual: 9
Predicted: 0, Actual: 0
Predicted: 1, Actual: 1
Predicted: 2, Actual: 2
Predicted: 3, Actual: 3
Predicted: 4, Actual: 4
Predicted: 5, Actual: 5
Predicted: 6, Actual: 6
Predicted: 7, Actual: 7
Predicted: 8, Actual: 8
Predicted: 9, Actual: 9
Predicted: 0, Actual: 0
Predicted: 1, Actual: 1
Predicted: 2, Actual: 2
Predicted: 3, Actual: 3
Predicted: 4, Actual: 4
Predicted: 5, Actual: 5
Predicted: 6, Actual: 6
Predicted: 7, Actual: 7
Predicted: 8, Actual: 8
Predicted: 9, Actual: 9
Predicted: 0, Actual: 0
Predicted: 9, Actual: 9
Predicted: 5, Actual: 5
Predicted: 5, Actual: 5
Predicted: 6, Actual: 6
Predicted: 5, Actual: 5
Predicted: 0, Actual: 0
Predicted: 9, Actual: 9
Predicted: 8, Actual: 8
Predicted: 9, Actual: 9
Predicted: 8, Actual: 8
Predicted: 4, Ac

In [12]:
# Display the score
# The score is the accuracy of the model
# The score is the percentage of correct predictions
# The score is the ratio of correct predictions to total predictions
model.score(X, y)

1.0

In [44]:
from pickle import dump, load

! dir

 Volume in drive C has no label.
 Volume Serial Number is DA42-B682

 Directory of c:\Users\maprado\Desktop\UCM\machinelearning3

05/11/2025  05:31 PM    <DIR>          .
05/09/2025  07:16 PM    <DIR>          ..
05/12/2025  08:49 AM           398,779 01-intro.ipynb
05/12/2025  08:14 AM             6,829 modelo-lr-001.pkl
               2 File(s)        405,608 bytes
               2 Dir(s)  182,241,099,776 bytes free


In [14]:
# Save the model to a file

# Save the model
with open("modelo-lr-001.pkl", "wb") as f:
    dump(model, f)

# Load the model
with open("modelo-lr-001.pkl", "rb") as f:
    modelolr001 = load(f)

In [15]:
# Display the saved model created using pickle
modelolr001

LogisticRegression()

In [16]:
modelolr001.get_params()

{'C': 1.0,
 'class_weight': None,
 'dual': False,
 'fit_intercept': True,
 'intercept_scaling': 1,
 'l1_ratio': None,
 'max_iter': 100,
 'multi_class': 'deprecated',
 'n_jobs': None,
 'penalty': 'l2',
 'random_state': None,
 'solver': 'lbfgs',
 'tol': 0.0001,
 'verbose': 0,
 'warm_start': False}

In [17]:
modelolr001.coef_

array([[ 0.00000000e+00, -5.45643843e-03, -4.12130745e-02,
         9.47979190e-02,  1.14111634e-02, -9.95292924e-02,
        -1.21942291e-01, -1.19397761e-02, -1.53988927e-04,
        -5.57845896e-02,  3.03414665e-02,  1.54087261e-01,
         8.81995241e-02,  1.30807851e-01, -5.13494588e-02,
        -1.20055567e-02, -1.72062022e-04,  4.15469140e-02,
         2.00672409e-01, -3.70270089e-02, -3.59401940e-01,
         1.96910240e-01,  8.57320712e-02, -4.12225049e-03,
        -7.83365223e-05,  1.53693154e-01,  1.26112095e-01,
        -2.41707620e-01, -5.44794696e-01,  2.87935599e-02,
         9.53895438e-02, -2.73015196e-04,  0.00000000e+00,
         1.87455501e-01,  1.20603629e-01, -2.52209638e-01,
        -4.77603514e-01,  4.12647203e-02,  6.75424465e-02,
         0.00000000e+00, -6.70032943e-04, -2.59723039e-02,
         3.00166268e-01, -1.75567748e-01, -1.57136766e-01,
         1.13928145e-01,  3.51638304e-02, -7.55014366e-04,
        -1.90357874e-03, -1.00859842e-01,  1.23677952e-0

## Modelo 2 - Mejorando el modelo

[Scikit Learn - train_test_split](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html)

In [18]:
from sklearn.model_selection import train_test_split

In [19]:
# Create a train/test split
# The train/test split is used to evaluate the model
# The train/test split is used to test the model on unseen data
# The train/test split is used to evaluate the performance of the model
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2
)

In [20]:
X_train

pixel_0_0  pixel_0_1  pixel_0_2  pixel_0_3  pixel_0_4  pixel_0_5  \
1190        0.0        3.0       16.0       13.0       15.0       16.0   
70          0.0        0.0        0.0        0.0       14.0        7.0   
1578        0.0        0.0        5.0       15.0       14.0       13.0   
985         0.0        0.0        3.0       15.0       16.0       12.0   
752         0.0        0.0        0.0        5.0       11.0        0.0   
...         ...        ...        ...        ...        ...        ...   
23          0.0        1.0        8.0       12.0       15.0       14.0   
535         0.0        0.0        1.0       12.0       16.0       14.0   
1532        0.0        0.0       15.0       14.0       15.0        9.0   
1282        0.0        0.0        9.0       16.0       16.0       10.0   
741         0.0        0.0        3.0       11.0        0.0        0.0   

      pixel_0_6  pixel_0_7  pixel_1_0  pixel_1_1  ...  pixel_6_6  pixel_6_7  \
1190       11.0        0.0        0.0        5.0  ...        0.0        0.0   
70          0.0        0.0        0.0        0.0  ...        0.0        0.0   
1578        2.0        0.0        0.0        0.0  ...        0.0        0.0   
985         0.0        0.0        0.0        0.0  ...       16.0        1.0   
752         0.0        0.0        0.0        0.0  ...       16.0        2.0   
...         ...        ...        ...        ...  ...        ...        ...   
23          4.0        0.0        0.0        3.0  ...        0.0        0.0   
535         2.0        0.0        0.0        0.0  ...        4.0        0.0   
1532        0.0        0.0        0.0        0.0  ...        1.0        0.0   
1282        0.0        0.0        0.0        2.0  ...        6.0        0.0   
741         0.0        0.0        0.0        0.0  ...       14.0        0.0   

      pixel_7_0  pixel_7_1  pixel_7_2  pixel_7_3  pixel_7_4  pixel_7_5  \
1190        0.0        2.0       15.0       15.0        9.0        0.0   
70          0.0        0.0        0.0        0.0       12.0       11.0   
1578        0.0        0.0        7.0       14.0        0.0        0.0   
985         0.0        0.0        3.0       14.0       16.0       16.0   
752         0.0        0.0        1.0        8.0       16.0       13.0   
...         ...        ...        ...        ...        ...        ...   
23          0.0        0.0       14.0       15.0       11.0        2.0   
535         0.0        0.0        3.0       12.0       16.0       11.0   
1532        0.0        1.0       10.0       16.0       15.0        4.0   
1282        0.0        0.0        9.0       16.0       16.0       11.0   
741         0.0        0.0        4.0       14.0       16.0       14.0   

      pixel_7_6  pixel_7_7  
1190        0.0        0.0  
70          0.0        0.0  
1578        0.0        0.0  
985         9.0        0.0  
752         9.0        1.0  
...         ...        ...  
23          0.0        0.0  
535         1.0        0.0  
1532        0.0        0.0  
1282        1.0        0.0  
741         7.0        0.0  

[1437 rows x 64 columns]

In [21]:
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((1437, 64), (360, 64), (1437,), (360,))

In [22]:
# Create a Logistic regression model
model2 = LogisticRegression()

# Fit the model
# The model is fit to the training data
# The model learns the relationship between X_train and y_train
model2.fit(X_train, y_train)

# Test the model
# The model is tested on the test data
# The model predicts the target values for the test data
# The model predicts the target values for X_test
model2.score(X_test, y_test)

0.9694444444444444

## Modelo 3 - Pipelines

* [Scitkit Learn - StandardScaler](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html)
* [Scitkit Learn - Pipeline](https://scikit-learn.org/stable/api/sklearn.pipeline.html)
  * [Pipeline](https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html)
  * [Make_pipeline](https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.make_pipeline.html#sklearn.pipeline.make_pipeline)
  * [Additional Docs - Apache Sparks Pipelines](https://spark.apache.org/docs/latest/ml-pipeline.html)

### Make_Pipeline

In [23]:
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline, Pipeline

In [24]:
# Create a train/test split
# The train/test split is used to evaluate the model
# The train/test split is used to test the model on unseen data
# The train/test split is used to evaluate the performance of the model
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2
)

# Fit the scaler to the training data
# The scaler learns the mean and standard deviation of the training data
# The scaler is used to transform the training data
scaler = StandardScaler()

# Select an algorithm
alg = LogisticRegression()

# Create a pipeline
# The pipeline is used to chain the scaler and the algorithm together
# The pipeline is used to fit the scaler and the algorithm to the training data
# The pipeline is used to transform the training data and fit the algorithm
model3 = make_pipeline(scaler, alg)

In [25]:
# Fit the model
# The model is fit to the training data
# The model learns the relationship between X_train and y_train
# The model is trained on the data
model3.fit(X_train, y_train)

# Test the model
# The model is tested on the test data
# The model predicts the target values for the test data
# The model predicts the target values for X_test
model3.score(X_test, y_test)

0.9638888888888889

In [26]:
#Display the model pipeline
model3

Pipeline(steps=[('standardscaler', StandardScaler()),
                ('logisticregression', LogisticRegression())])

In [27]:
# Display the model steps
# The model steps are the scaler and the algorithm
# The model steps are the components of the pipeline
# The model steps are the scaler and the algorithm
model3.steps

[('standardscaler', StandardScaler()),
 ('logisticregression', LogisticRegression())]

In [28]:
# Display the model step -1
# The model step -1 is the Logistic regression algorithm
# The model step -1 is the algorithm that is used to make predictions
model3.steps[-1][1]

LogisticRegression()

In [29]:
# Display the Logistic regression coefficients
# The coefficients are the weights of the model
# The coefficients are the values that the model learned
# The coefficients are used to make predictions
model3.steps[-1][1].coef_

array([[ 0.00000000e+00,  3.23980378e-02, -9.51086947e-02,
         1.29519767e-01, -3.79227924e-02, -4.39844059e-01,
        -2.98771040e-01, -6.94263465e-02, -1.71135133e-02,
        -1.12434910e-02, -1.77773586e-01,  3.38057062e-01,
         1.04707423e-01,  9.05127043e-02, -1.64845409e-01,
        -9.90343133e-02, -1.04736641e-02,  6.30219878e-02,
         1.45426682e-01, -1.15930632e-01, -6.63458013e-01,
         4.17515830e-01,  1.94067282e-01, -4.20330440e-02,
        -1.17388831e-02,  2.01275257e-01,  2.47817003e-01,
        -5.24512857e-01, -1.14381416e+00,  5.58175423e-02,
         4.28990657e-01, -2.19874856e-03,  0.00000000e+00,
         3.80185661e-01,  2.91223164e-01, -5.90545761e-01,
        -9.98894244e-01,  6.39582950e-02,  2.47380952e-01,
         0.00000000e+00, -4.08339124e-03,  5.82007343e-02,
         6.00759346e-01, -3.80052120e-01, -2.91921121e-01,
         1.39214673e-01,  1.16018019e-01, -8.24463550e-03,
        -2.41732467e-03, -1.34638134e-01,  2.89479414e-0

In [30]:
# Display the StandardScaler mean
# The mean is the average of the training data
# The mean is used to center the data
# The mean is used to transform the training data
model3.steps[0][1].mean_

array([0.00000000e+00, 2.98538622e-01, 5.14196242e+00, 1.18114127e+01,
       1.18733473e+01, 5.76757133e+00, 1.37299930e+00, 1.37787056e-01,
       6.26304802e-03, 1.95755045e+00, 1.03075852e+01, 1.19832985e+01,
       1.02254697e+01, 8.14961726e+00, 1.85038274e+00, 1.13430759e-01,
       2.78357690e-03, 2.55114823e+00, 9.93110647e+00, 7.10508003e+00,
       7.06332637e+00, 7.78427279e+00, 1.78844816e+00, 5.01043841e-02,
       6.95894224e-04, 2.44954767e+00, 9.14752958e+00, 8.86638831e+00,
       9.84899095e+00, 7.56576200e+00, 2.35560195e+00, 2.78357690e-03,
       0.00000000e+00, 2.34446764e+00, 7.74947808e+00, 9.17884482e+00,
       1.03639527e+01, 8.75922060e+00, 2.96520529e+00, 0.00000000e+00,
       9.74251914e-03, 1.61864997e+00, 6.94502436e+00, 7.37091162e+00,
       7.72720946e+00, 8.21224774e+00, 3.39874739e+00, 2.78357690e-02,
       9.04662491e-03, 7.29297147e-01, 7.52957550e+00, 9.63952679e+00,
       9.37369520e+00, 8.68754349e+00, 3.72651357e+00, 2.21990257e-01,
      

In [31]:
# Display the StandardScaler variance
# The variance is the spread of the training data
# The variance is used to scale the data
# The variance is used to transform the training data
model3.steps[0][1].var_

array([0.00000000e+00, 8.09274134e-01, 2.24224354e+01, 1.81502386e+01,
       1.84014956e+01, 3.22159839e+01, 1.12818875e+01, 1.14454987e+00,
       1.03991876e-02, 9.98915141e+00, 2.98747720e+01, 1.59273481e+01,
       2.23555656e+01, 3.71613308e+01, 1.28620962e+01, 7.33827965e-01,
       4.16761704e-03, 1.26315175e+01, 3.22659565e+01, 3.38602178e+01,
       3.78797754e+01, 3.87579155e+01, 1.06928937e+01, 1.93731722e-01,
       6.95409955e-04, 9.84661949e+00, 3.77110116e+01, 3.45402551e+01,
       3.78080940e+01, 3.45449099e+01, 1.39159968e+01, 2.77582860e-03,
       0.00000000e+00, 1.21951903e+01, 3.97298623e+01, 3.92074021e+01,
       3.52294034e+01, 3.42343009e+01, 1.26153516e+01, 0.00000000e+00,
       2.35654869e-02, 9.13501752e+00, 4.29301718e+01, 4.12910954e+01,
       3.93563438e+01, 3.28324735e+01, 1.85515085e+01, 8.96914191e-02,
       5.21102254e-02, 3.21760375e+00, 3.23660355e+01, 2.73648399e+01,
       2.81867263e+01, 3.66226215e+01, 2.46134446e+01, 1.02170154e+00,
      

### Pipeline

In [32]:
# Define the pipeline
model3_b = Pipeline([
    ('estandar', StandardScaler()),
    ('LR', LogisticRegression())
])

#Display the pipeline
model3_b

Pipeline(steps=[('estandar', StandardScaler()), ('LR', LogisticRegression())])

In [33]:
# Display the pipeline steps
# The pipeline steps are the scaler and the algorithm
# The pipeline steps are the components of the pipeline
# The pipeline steps are the scaler and the algorithm
model3_b.steps

[('estandar', StandardScaler()), ('LR', LogisticRegression())]

### Pipeline + custom arguments

In [34]:
# Define the pipeline
model3_c = Pipeline([
    ('estandar', StandardScaler()),
    ('LR', LogisticRegression(solver='newton-cholesky'))
])

#Display the pipeline
model3_c

Pipeline(steps=[('estandar', StandardScaler()),
                ('LR', LogisticRegression(solver='newton-cholesky'))])

## Comparacion de modelos

* [Scikit Learn - DecisionTreeClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html)
* [Scikit Learn - Gaussian Naive Bayes](https://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.GaussianNB.html)
* [Scikit Learn - K Neighbors Classifier](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html)
* [Scikit Learn - Linear Discriminant Analysis](https://scikit-learn.org/stable/modules/generated/sklearn.discriminant_analysis.LinearDiscriminantAnalysis.html)
* [Scikit Learn - Random Forest Classifier](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html)
* [Scikit Learn - ](https://scikit-learn.org/stable/modules/generated/sklearn.svm.LinearSVC.html)

In [ ]:
# Libraries
from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.neighbors import KNeighborsClassifier
from  sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import LinearSVC

In [40]:
seed = 99

In [ ]:
# Define the models in a list
models = []
models.append(('LR', LogisticRegression(random_state=seed)))
models.append(('DTC', DecisionTreeClassifier(random_state=seed)))
models.append(('GNB', GaussianNB()))
models.append(('KNN', KNeighborsClassifier()))
models.append(('LDA', LinearDiscriminantAnalysis()))
models.append(('RFC', RandomForestClassifier(random_state=seed)))
models.append(('SVM', LinearSVC(random_state=seed)))

In [ ]:
# Display the models
models

[('LR', LogisticRegression(random_state=99)),
 ('DTC', DecisionTreeClassifier(random_state=99)),
 ('GNB', GaussianNB()),
 ('KNN', KNeighborsClassifier()),
 ('LDA', LinearDiscriminantAnalysis()),
 ('RFC', RandomForestClassifier(random_state=99)),
 ('SVM', LinearSVC(random_state=99))]

In [43]:
for name, model in models:
    # Fit the model
    # The model is fit to the training data
    # The model learns the relationship between X_train and y_train
    # The model is trained on the data
    model.fit(X_train, y_train)
    acc = model.score(X_test, y_test)
    print(f"Model {name} Accuracy: {acc:.2f}")


Model LR Accuracy: 0.96
Model DTC Accuracy: 0.84
Model GNB Accuracy: 0.83
Model KNN Accuracy: 0.99
Model LDA Accuracy: 0.95
Model RFC Accuracy: 0.98
Model SVM Accuracy: 0.93


### Model Performance Summary and Ranking

The following are the accuracy scores for each model evaluated on the test dataset:

| Rank | Model Name                      | Accuracy |
|------|---------------------------------|----------|
| 1    | K-Nearest Neighbors (KNN)       | 0.99     |
| 2    | Random Forest Classifier (RFC)  | 0.98     |
| 3    | Logistic Regression (LR)        | 0.96     |
| 4    | Linear Discriminant Analysis (LDA)| 0.95   |
| 5    | Support Vector Machine (SVM)    | 0.93     |
| 6    | Decision Tree Classifier (DTC)  | 0.84     |
| 7    | Gaussian Naive Bayes (GNB)      | 0.83     |

### Interpretation:

- **K-Nearest Neighbors (KNN)** achieved the highest accuracy (0.99), indicating excellent predictive performance on this dataset.
- **Random Forest Classifier (RFC)** also performed very well (0.98), closely following KNN.
- **Logistic Regression (LR)** and **Linear Discriminant Analysis (LDA)** provided strong results, with accuracies of 0.96 and 0.95 respectively.
- **Support Vector Machine (SVM)** showed moderate performance (0.93).
- **Decision Tree Classifier (DTC)** and **Gaussian Naive Bayes (GNB)** had the lowest accuracies (0.84 and 0.83 respectively), indicating weaker predictive performance compared to the other models.

Based on these results, **KNN** is the best-performing model, while **Gaussian Naive Bayes** is the least effective for this dataset.